tests/unit/test_zookeeper_provider.py

```diff
@@ -0,0 +1,291 @@
+        )
+
+        self.assertEqual(config_values, {"relation-2", "relation-0"})
+
+    def test_is_child_of(self):
+        chroot = "/gandalf/the/white"
+        chroots = {"/gandalf", "/saruman"}
+
+        self.assertTrue(
+            self.harness.charm.client_relation._is_child_of(path=chroot, chroots=chroots)
+        )
+
+    def test_apply_relation_data(self):
+        self.harness.set_leader(True)
+        self.harness.add_relation("database", "new_application")
+        self.harness.update_relation_data(
+            self.provider.client_relations[0].id, "application", {"chroot": "app"}
+        )
+        self.harness.update_relation_data(
+            self.provider.client_relations[1].id,
+            "new_application",
+            {"chroot": "new_app", "chroot-acl": "rw"},
+        )
+        self.harness.update_relation_data(
+            self.provider.app_relation.id,
+            "zookeeper/0",
+            {"private-address": "treebeard", "state": "started"},
+        )
+        self.harness.add_relation_unit(self.provider.app_relation.id, "zookeeper/1")
+        self.harness.update_relation_data(
+            self.provider.app_relation.id,
+            "zookeeper/1",
+            {"private-address": "shelob", "state": "ready"},
+        )
+        self.harness.add_relation_unit(self.provider.app_relation.id, "zookeeper/2")
+        self.harness.update_relation_data(
+            self.provider.app_relation.id,
+            "zookeeper/2",
+            {"private-address": "balrog", "state": "started"},
+        )
+
+        self.harness.charm.client_relation.apply_relation_data()
+
+        passwords = []
+        usernames = []
+        for relation in self.provider.client_relations:
+
+            # checking existence of all necessary keys
+            self.assertEqual(
+                sorted(relation.data[self.harness.charm.app].keys()),
+                sorted(["chroot", "endpoints", "password", "uris", "username"]),
+            )
+
+            # checking unique passwords and usernames for all relations
+            self.assertNotIn(relation.data[self.harness.charm.app]["password"], passwords)
+            self.assertNotIn(relation.data[self.harness.charm.app]["username"], usernames)
+
+            # checking multiple endpoints and uris
+            self.assertEqual(len(relation.data[self.harness.charm.app]["endpoints"].split(",")), 2)
+            self.assertEqual(len(relation.data[self.harness.charm.app]["uris"].split(",")), 2)
+
+            for uri in relation.data[self.harness.charm.app]["uris"].split(","):
+                # checking chroot in uri
+                self.assertTrue(uri.endswith(relation.data[self.harness.charm.app]["chroot"]))
+                # checking client_port in uri
+                self.assertTrue(re.search(r":[\d]+\/", uri))
+
+            passwords.append(relation.data[self.harness.charm.app]["password"])
+            usernames.append(relation.data[self.harness.charm.app]["username"])
```

Comment:
It's be interesting to also check the app relation (peer) data, as it's being updated in the method that is being tested.